### PR TITLE
POC: Short circuit render path for streaming data frames

### PR DESCRIPTION
**Pull request by ryantxu**
**What this PR does / why we need it**:

This adds a "appendObserver" to the DataFrame interface that can let panels subscribe directly to append only events.  This is an effort to combat some serious CPU issues with streaming data sources.

In this PR, I explored a fast render path with a super simple panel (it just renders the data frame as json string).  This is the approximate timing differences: (all values are ~CPU usage from the chrome "Performance Monitor" on my laptop)
image

The most obvious and glaring fact is that CPU behavior on the existing edit view is dramatically higher than it should be.  Fixing that is likely possible even without this strategy.

Peek 2021-03-14 12-35

NOTE: this is testing with the dev builds, production builds are a bit better, but the relative values are similar.  In production the 20hz observer is ~15% (vs 18)

### CODE REVIEW

File: public/app/plugins/panel/stat/StatPanel.tsx [hunk 1]
```diff
@@ -8,18 +8,56 @@ import {
...
+
+  componentDidMount() {}
+
+  componentDidUpdate(oldProps: Props) {
+    const { data } = this.props;
+    if (oldProps.data !== data) {
+      for (const series of data.series) {
+        if (series.appendObserver) {
+          //console.log('TODO subscribe!!!!', series);
+          series.appendObserver.subscribe(this.events);
```
Comment (ryantxu) [inline public/app/plugins/panel/stat/StatPanel.tsx:42]: to toggle "current" vs "observer" behavior, I just commented out this line

**Comment by torkelo** [comment]
For edit view it's tricky and simple I think, it's tricky in that both query editors & panel options need panel data, and re-render all those components is not super cheap. But I think the simple & quick solution should be just to throttle the data update to query editors and the options pane

**Review by torkelo — COMMENTED** [review]
The biggest problem with this is that it will have to be implemented in each panel, but I guess it's mainly timeseries that will need this

**Comment by ryantxu** [comment]
> implemented in each panel

Right -- that is both the problem and advantage.  This has an explicit observer to *skip* the normal react paths -- I will also explore adding a special key to frame so we can easily tell (without calculating) that the packet is a simple update/append rather than something that requires a full refresh

**Comment by ryantxu** [comment]
> I think the simple & quick solution should be just to throttle

That will help us stay in a usable zone +1, but we may *also* want to check if the schema changes.  none of our current editors care about the field values, only the names/types

**Comment by leeoniya** [comment]
i profiled this with `display: none` applied to the `pre` div to eliminate style/reflow costs that dominate the measurements. in production builds this dropped my CPU to 4% for the standard path and 1.5-2% for the observable path, so about a 50-60% overhead reduction. given that the numbers are single-digit %, it may not seem terribly impactful in this basic bench, but this is for a panel with 3 fields, no custom config or overrides, no transforms. the effect on more realistic panels should be much greater. even ryan's numbers with just the vis enabled already show gains of not the 2% saved overhead, but 7%. looking forward to try this out in timeseries panels.

**Comment by torkelo** [comment]
closing this as we are going down a different path

**Pull request closed without merging**